src/Symfony/Component/Semaphore/Store/RedisStore.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Semaphore\Store;
+
+use Symfony\Component\Cache\Traits\RedisClusterProxy;
+use Symfony\Component\Cache\Traits\RedisProxy;
+use Symfony\Component\Semaphore\Exception\InvalidArgumentException;
+use Symfony\Component\Semaphore\Exception\SemaphoreAcquiringException;
+use Symfony\Component\Semaphore\Exception\SemaphoreExpiredException;
+use Symfony\Component\Semaphore\Key;
+use Symfony\Component\Semaphore\PersistingStoreInterface;
+
+/**
+ * RedisStore is a PersistingStoreInterface implementation using Redis as store engine.
+ *
+ * @experimental in 5.1
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RedisStore implements PersistingStoreInterface
+{
+    private $redis;
+
+    /**
+     * @param \Redis|\RedisArray|\RedisCluster|\Predis\ClientInterface $redisClient
+     */
+    public function __construct($redisClient)
+    {
+        if (!$redisClient instanceof \Redis && !$redisClient instanceof \RedisArray && !$redisClient instanceof \RedisCluster && !$redisClient instanceof \Predis\ClientInterface && !$redisClient instanceof RedisProxy) {
+            throw new InvalidArgumentException(sprintf('"%s()" expects parameter 1 to be Redis, RedisArray, RedisCluster or Predis\ClientInterface, "%s" given.', __METHOD__, \is_object($redisClient) ? \get_class($redisClient) : \gettype($redisClient)));
+        }
+
+        $this->redis = $redisClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key, float $ttlInSecond)
+    {
+        if (0 > $ttlInSecond) {
+            throw new InvalidArgumentException("The TTL should be greater than 0, '$ttlInSecond' given.");
+        }
+
+        $script = file_get_contents(__DIR__.'/scripts/redis_save.lua');
```

Comment:
could be `static $script` or `$script = include(...);`to reduce I/O

Reply:
I'm not sure it will reduce IO: the file will be in RAM (cache by the OS layer) and maybe PHP does it too ?!

Reply:
I tested it with blackfire:

https://blackfire.io/profiles/2759c130-2a27-4042-a128-ebaffb9bc82e/graph?settings%5Bdimension%5D=io&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=&callname=main()

it's so small we don't even see it

The code:
```php

$redis = new Redis();
$redis->connect('172.17.0.2');

$semaphore = (new SemaphoreFactory(new RedisStore($redis)))->createSemaphore('test', 1, 1, 10);


for ($i=0; $i < 10000; $i++) {
    if (!$semaphore->acquire()) {
    }
}
```

And I ran another process that already acquired the only slot available.


---

I prefer to avoid this premature optimisation. But thanks for the tips